Spigot-Server-Patches/0622-Strip-raytracing-for-EntityLiving-hasLineOfSight.patch

```diff
@@ -0,0 +1,136 @@
+From: Paul Sauve <paul@technove.co>
+Date: Sat, 31 Oct 2020 18:43:02 -0500
+Subject: [PATCH] Strip raytracing for EntityLiving#hasLineOfSight
+
+The IBlockAccess#rayTrace method is very wasteful in both allocations,
+and in logic. While EntityLiving#hasLineOfSight provides static
+parameters for collisions with blocks and fluids, the method still does
+a lot of dynamic checks for both of these, which result in extra work.
+As well, since the fluid collision option is set to NONE, the entire
+fluid collision system is completely unneeded, yet used anyways.
+
+This patch was originally created for the AirplaneLite project (https://github.com/Technove/AirplaneLite) under the GPLv3 License.
+
+diff --git a/src/main/java/net/minecraft/server/EntityLiving.java b/src/main/java/net/minecraft/server/EntityLiving.java
+index 0294ff10dc8edb24aabfbe1589048a405c356ef0..e65f802f376f3c9debbd0ccc76e57a411c777405 100644
+--- a/src/main/java/net/minecraft/server/EntityLiving.java
++++ b/src/main/java/net/minecraft/server/EntityLiving.java
+@@ -2979,7 +2979,7 @@ public abstract class EntityLiving extends Entity {
+         Vec3D vec3d = new Vec3D(this.locX(), this.getHeadY(), this.locZ());
+         Vec3D vec3d1 = new Vec3D(entity.locX(), entity.getHeadY(), entity.locZ());
+ 
+-        return this.world.rayTrace(new RayTrace(vec3d, vec3d1, RayTrace.BlockCollisionOption.COLLIDER, RayTrace.FluidCollisionOption.NONE, this)).getType() == MovingObjectPosition.EnumMovingObjectType.MISS;
++        return this.world.rayTraceDirect(vec3d, vec3d1, VoxelShapeCollision.a(this)) == MovingObjectPosition.EnumMovingObjectType.MISS; // Paper - use direct method
+     }
+ 
+     @Override
+diff --git a/src/main/java/net/minecraft/server/IBlockAccess.java b/src/main/java/net/minecraft/server/IBlockAccess.java
+index c4a83448ed4513f6e4ab179d1d43e5bb0cb13641..62258e5d75d259664c107d78d12e7284830bbca2 100644
+--- a/src/main/java/net/minecraft/server/IBlockAccess.java
++++ b/src/main/java/net/minecraft/server/IBlockAccess.java
+@@ -44,6 +44,16 @@ public interface IBlockAccess {
+         return BlockPosition.a(axisalignedbb).map(this::getType);
+     }
+ 
++    // Paper start - broken down variant of below rayTraceBlock, used by World#rayTraceDirect
++    default MovingObjectPosition.EnumMovingObjectType rayTraceBlockDirect(Vec3D vec3d, Vec3D vec3d1, BlockPosition blockposition, IBlockData iblockdata, VoxelShapeCollision voxelshapecoll) {
++        RayTrace.BlockCollisionOption raytrace_blockcollisionoption = RayTrace.BlockCollisionOption.COLLIDER;
++        VoxelShape voxelshape = raytrace_blockcollisionoption.get(iblockdata, this, blockposition, voxelshapecoll); // Collider doesn't actually need a voxelshape
++        MovingObjectPositionBlock movingobjectpositionblock = this.rayTrace(vec3d, vec3d1, blockposition, voxelshape, iblockdata);
++
++        return movingobjectpositionblock == null ? null : movingobjectpositionblock.getType(); // can't be null because of the logic above
```

Comment:
What can't be null? You're doing a null-check, yet saying something can't be null where you return a nullable value.

Reply:
I think the reasoning was that in the previous code that could be null, but since the code that could make it null was stripped out the type can't be null.

Reply:
I'd say just drop the comment, then. No need making us confused :p

Reply:
I didn't submit this here, was more for personal reference if I come back to it in the future.

Reply:
fixed

Reply:
not fixed

Reply:
I misunderstood @Proximyst's comment lol